### ISSUE TITLE
firebase.auth().currentUser sometimes undefined

### ISSUE THREAD

**Issue by booboothefool** [invertase/react-native-firebase#3251]
## Issue

I intermittently get this error on about 5% of requests. It seems to work most of the time, but I'd like to get it to 100%. We discovered it is because `firebase.auth().currentUser` sometimes ends up `undefined` (but the user should be authenticated already) so the token passed is `Bearer undefined`. Is there a way to get `firebase.auth().currentUser` and the id token more reliably?

```
{"name":"myapp","hostname":"worker-844ddfbc9f-ntlmz","pid":18,"level":50,
"err":"[Throws: Cannot read property 'code' of undefined]",
"msg":"Context creation failed: Decoding Firebase ID token failed. Make sure you passed the entire string JWT which represents an ID token. See https://firebase.google.com/docs/auth/admin/verify-id-tokens for details on how to retrieve an ID token.",
"time":"2020-02-12T02:16:33.538Z","v":0}
TypeError: Cannot read property 'code' of undefined
    at FirebaseAuthError.get [as code] (/app/node_modules/firebase-admin/lib/utils/error.js:51:35)
    at FirebaseAuthError.FirebaseError.toJSON (/app/node_modules/firebase-admin/lib/utils/error.js:67:24)
    at JSON.stringify (<anonymous>)
    at prettyJSONStringify (/app/node_modules/apollo-server-core/dist/runHttpQuery.js:257:17)
    at throwHttpGraphQLError (/app/node_modules/apollo-server-core/dist/runHttpQuery.js:26:42)
    at Object.<anonymous> (/app/node_modules/apollo-server-core/dist/runHttpQuery.js:66:28)
    at Generator.next (<anonymous>)
    at fulfilled (/app/node_modules/apollo-server-core/dist/runHttpQuery.js:4:58)
    at process._tickCallback (internal/process/next_tick.js:68:7)
```

**client**
```
const authLink = setContext(async (_, { headers }) => {
    let token;
    if (firebase.auth().currentUser) {
        token = await firebase.auth().currentUser.getIdToken();
    }
    return {
        headers: {
            ...headers,
            authorization: `Bearer ${token}`,
        },
    };
});
```

**server**
```
            let token = req.headers.authorization;

            if (!token || !token.length) {
                console.log('no token');
            }

            token = token.split('Bearer ')[1];

            // validate JWT and pluck user id
            const { uid } = await firebase.auth().verifyIdToken(token);

            // find the user based on id
            const user = await firebase.auth().getUser(uid);
```

---

### Javascript

<details><summary>Click To Expand</summary>
<p>

#### `package.json`:

```json
{
  "name": "client",
  "version": "0.1.0",
  "private": true,
  "devDependencies": {
    "babel-loader": "^7.1.2",
    "babel-plugin-expo-web": "^0.0.5",
    "babel-plugin-react-native-web": "^0.4.0",
    "babel-plugin-transform-decorators-legacy": "^1.3.4",
    "babel-plugin-transform-imports": "^1.4.1",
    "babel-plugin-transform-runtime": "^6.23.0",
    "css-loader": "^0.28.7",
    "file-loader": "^1.1.7",
    "jest-expo": "~27.0.0",
    "react-native-scripts": "latest",
    "react-test-renderer": "16.3.1",
    "style-loader": "^0.19.0"
  },
  "main": "./node_modules/react-native-scripts/build/bin/crna-entry-web.js",
  "scripts": {
    "start": "expo start",
    "ios": "expo ios",
    "android": "expo android",
    "eject": "expo eject",
    "test": "jest",
    "web": "webpack-dev-server -d --config ./webpack.config.js  --inline --hot --colors --content-base public/ --history-api-fallback",
    "build": "NODE_ENV=production webpack -p --config ./webpack.config.js"
  },
  "jest": {
    "preset": "jest-expo"
  },
  "dependencies": {
    "@ptomasroos/react-native-multi-slider": "^0.0.14",
    "@react-native-community/async-storage": "^1.5.0",
    "apollo-boost": "^0.1.7",
    "apollo-cache-persist": "^0.1.1",
    "apollo-client": "2.4.2",
    "apollo-link-batch-http": "^1.2.8",
    "apollo-link-context": "^1.0.9",
    "apollo-link-http": "^1.5.4",
    "apollo-link-schema": "^1.1.0",
    "apollo-link-state": "^0.4.1",
    "apollo-link-ws": "^1.0.9",
    "apollo-utilities": "1.0.21",
    "axios": "^0.18.0",
    "babel-plugin-transform-remove-console": "^6.9.4",
    "expo": "30.0.0",
    "expo-web": "^0.0.12",
    "expokit": "1.7.1",
    "faker": "^4.1.0",
    "geolib": "^2.0.24",
    "graphql": "^0.13.2",
    "graphql-iso-date": "^3.5.0",
    "graphql-tag": "^2.9.2",
    "graphql-tools": "^3.0.2",
    "immutability-helper": "^3.0.0",
    "qs": "^6.7.0",
    "react": "16.3.1",
    "react-apollo": "^2.1.4",
    "react-dom": "16.0.0",
    "react-native": "https://github.com/expo/react-native/archive/sdk-30.0.0.tar.gz",
    "react-native-animatable": "^1.3.1",
    "react-native-check-box": "^2.1.0",
    "react-native-circular-progress": "^1.1.0",
    "react-native-collapsible": "^0.11.3",
    "react-native-confirmation-code-field": "3.7.0",
    "react-native-country-picker-modal": "^0.8.0",
    "react-native-datepicker": "^1.7.2",
    "react-native-deck-swiper": "^1.5.15",
    "react-native-device-info": "^1.6.1",
    "react-native-dropdownalert": "^4.1.0",
    "react-native-fast-image": "^5.1.1",
    "react-native-fbsdk": "^0.8.0",
    "react-native-firebase": "^5.0.0-rc4",
    "react-native-gifted-chat": "^0.4.3",
    "react-native-haptic-feedback": "^1.6.0",
    "react-native-hyperlink": "^0.0.14",
    "react-native-iap": "2.3.26",
    "react-native-iphone-x-helper": "^1.2.0",
    "react-native-loading-spinner-overlay": "^1.0.1",
    "react-native-modal": "^11.0.1",
    "react-native-onboarding-swiper": "^1.0.0",
    "react-native-parallax-scroll-view": "^0.21.3",
    "react-native-phone-input": "^0.2.4",
    "react-native-simple-radio-button": "^2.7.3",
    "react-native-smart-corner-label": "^1.1.2",
    "react-native-smart-timer-enhance": "^1.0.3",
    "react-native-snap-carousel": "^3.8.1",
    "react-native-sortable-listview": "^0.2.8",
    "react-native-storage": "^1.0.1",
    "react-native-svg-animated-linear-gradient": "^0.1.9",
    "react-native-swipe-list-view": "^1.5.1",
    "react-native-swiper": "^1.5.13",
    "react-native-tab-view": "^1.3.2",
    "react-native-vector-icons": "^6.3.0",
    "react-native-web": "^0.7.3",
    "react-navigation": "^3.3.0",
    "react-navigation-fluid-transitions": "^0.2.74",
    "react-primitives": "^0.5.1",
    "rn-placeholder": "^2.0.0",
    "styled-components": "^3.2.6",
    "subscriptions-transport-ws": "^0.9.15",
    "uuid": "^3.3.2",
    "webpack": "^3.11.0",
    "webpack-dev-server": "2.9.4"
  }
}

```

#### `firebase.json` for react-native-firebase v6:

```json
# N/A
```

</p>
</details>

---

### iOS

<details><summary>Click To Expand</summary>
<p>

#### `ios/Podfile`:

- [ ] I'm not using Pods
- [x] I'm using Pods and my Podfile looks like:

```ruby
source 'https://github.com/CocoaPods/Specs.git'
platform :ios, '9.0'

target 'client' do
  pod 'ExpoKit',
    :git => "http://github.com/expo/expo.git",
    :tag => "ios/2.8.3",
    :subspecs => [
      "Core"
    ],
    :inhibit_warnings => true
  pod 'EXAdsAdMob',
    :path => "../node_modules/expo-ads-admob/ios"
  pod 'EXBarCodeScanner',
    :path => "../node_modules/expo-barcode-scanner/ios"
  pod 'EXBarCodeScannerInterface',
    :path => "../node_modules/expo-barcode-scanner-interface/ios"
  pod 'EXCamera',
    :path => "../node_modules/expo-camera/ios"
  pod 'EXCameraInterface',
    :path => "../node_modules/expo-camera-interface/ios"
  pod 'EXConstants',
    :path => "../node_modules/expo-constants/ios"
  pod 'EXConstantsInterface',
    :path => "../node_modules/expo-constants-interface/ios"
  pod 'EXContacts',
    :path => "../node_modules/expo-contacts/ios"
  pod 'EXCore',
    :path => "../node_modules/expo-core/ios"
  pod 'EXFaceDetectorInterface',
    :path => "../node_modules/expo-face-detector-interface/ios"
  pod 'EXFileSystem',
    :path => "../node_modules/expo-file-system/ios"
  pod 'EXFileSystemInterface',
    :path => "../node_modules/expo-file-system-interface/ios"
  pod 'EXFont',
    :path => "../node_modules/expo-font/ios"
  pod 'EXFontInterface',
    :path => "../node_modules/expo-font-interface/ios"
  pod 'EXGL',
    :path => "../node_modules/expo-gl/ios"
  pod 'EXGL-CPP',
    :path => "../node_modules/expo-gl-cpp/cpp"
  pod 'EXImageLoaderInterface',
    :path => "../node_modules/expo-image-loader-interface/ios"
  pod 'EXLocalAuthentication',
    :path => "../node_modules/expo-local-authentication/ios"
  pod 'EXLocation',
    :path => "../node_modules/expo-location/ios"
  pod 'EXMediaLibrary',
    :path => "../node_modules/expo-media-library/ios"
  pod 'EXPermissions',
    :path => "../node_modules/expo-permissions/ios"
  pod 'EXPermissionsInterface',
    :path => "../node_modules/expo-permissions-interface/ios"
  pod 'EXPrint',
    :path => "../node_modules/expo-print/ios"
  pod 'EXReactNativeAdapter',
    :path => "../node_modules/expo-react-native-adapter/ios"
  pod 'EXSegment',
    :path => "../node_modules/expo-analytics-segment/ios"
  pod 'EXSensors',
    :path => "../node_modules/expo-sensors/ios"
  pod 'EXSensorsInterface',
    :path => "../node_modules/expo-sensors-interface/ios"
  pod 'EXSMS',
    :path => "../node_modules/expo-sms/ios"

  pod 'React',
    :path => "../node_modules/react-native",
    :inhibit_warnings => true,
    :subspecs => [
      "Core",
      "ART",
      "RCTActionSheet",
      "RCTAnimation",
      "RCTCameraRoll",
      "RCTGeolocation",
      "RCTImage",
      "RCTNetwork",
      "RCTText",
      "RCTVibration",
      "RCTWebSocket",
      "DevSupport",
      "CxxBridge"
    ]
  pod 'yoga',
    :path => "../node_modules/react-native/ReactCommon/yoga",
    :inhibit_warnings => true
  pod 'DoubleConversion',
    :podspec => "../node_modules/react-native/third-party-podspecs/DoubleConversion.podspec",
    :inhibit_warnings => true
  pod 'Folly',
    :podspec => "../node_modules/react-native/third-party-podspecs/Folly.podspec",
    :inhibit_warnings => true
  pod 'glog',
    :podspec => "../node_modules/react-native/third-party-podspecs/glog.podspec",
    :inhibit_warnings => true
  
  pod 'Firebase/Core', '~> 5.3.0'
  pod 'Firebase/Auth', '~> 5.3.0'
  pod 'Firebase/Messaging', '~> 5.3.0'
  pod 'Firebase/Storage', '~> 5.3.0'
  pod 'Firebase/DynamicLinks', '~> 5.3.0'

  pod 'react-native-fast-image', :path => '../node_modules/react-native-fast-image'

  pod 'RNVectorIcons', :path => '../node_modules/react-native-vector-icons'

  pod 'react-native-haptic-feedback', :path => '../node_modules/react-native-haptic-feedback'

  pod 'react-native-fbsdk', :path => '../node_modules/react-native-fbsdk'

  pod 'RNDeviceInfo', :path => '../node_modules/react-native-device-info'

  pod 'RNCAsyncStorage', :path => '../node_modules/@react-native-community/async-storage'

  post_install do |installer|
    installer.pods_project.main_group.tab_width = '2';
    installer.pods_project.main_group.indent_width = '2';

    installer.pod_targets.each do |target|

    if target.pod_name == 'ExpoKit'
      target.native_target.build_configurations.each do |config|
        config.build_settings['GCC_PREPROCESSOR_DEFINITIONS'] ||= ['$(inherited)']
        config.build_settings['GCC_PREPROCESSOR_DEFINITIONS'] << 'EX_DETACHED=1'
        
        # needed for GoogleMaps 2.x
        config.build_settings['FRAMEWORK_SEARCH_PATHS'] ||= []
        config.build_settings['FRAMEWORK_SEARCH_PATHS'] << '${PODS_ROOT}/GoogleMaps/Base/Frameworks'
        config.build_settings['FRAMEWORK_SEARCH_PATHS'] << '${PODS_ROOT}/GoogleMaps/Maps/Frameworks'
      end
    end


    if ['Amplitude-iOS','Analytics','AppAuth','Branch','CocoaLumberjack','FBSDKCoreKit','FBSDKLoginKit','FBSDKShareKit','GPUImage','JKBigInteger2'].include? target.pod_name
      target.native_target.build_configurations.each do |config|
        config.build_settings['IPHONEOS_DEPLOYMENT_TARGET'] = '9.0'
      end
    end
    # Can't specify this in the React podspec because we need
    # to use those podspecs for detached projects which don't reference ExponentCPP.
    if target.pod_name.start_with?('React')
      target.native_target.build_configurations.each do |config|
        config.build_settings['IPHONEOS_DEPLOYMENT_TARGET'] = '9.0'
        config.build_settings['HEADER_SEARCH_PATHS'] ||= ['$(inherited)']
      end
    end
    # Build React Native with RCT_DEV enabled
    next unless target.pod_name == 'React'
    target.native_target.build_configurations.each do |config|
      config.build_settings['GCC_PREPROCESSOR_DEFINITIONS'] ||= ['$(inherited)']
      config.build_settings['GCC_PREPROCESSOR_DEFINITIONS'] << 'RCT_DEV=1'
    end

    end
  end
end

```

#### `AppDelegate.m`:

```objc
// Copyright 2015-present 650 Industries. All rights reserved.

#import "AppDelegate.h"
#import "ExpoKit.h"
#import "EXViewController.h"

#import <Firebase.h>
#import "RNFirebaseNotifications.h"
#import "RNFirebaseMessaging.h"
#import <FBSDKCoreKit/FBSDKCoreKit.h>

@interface AppDelegate ()

@property (nonatomic, strong) EXViewController *rootViewController;

@end

@implementation AppDelegate

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
{
    [FIRApp configure];
    [RNFirebaseNotifications configure];
    [[FBSDKApplicationDelegate sharedInstance] application:application
                             didFinishLaunchingWithOptions:launchOptions];

    _window = [[UIWindow alloc] initWithFrame:[UIScreen mainScreen].bounds];
    _window.backgroundColor = [UIColor whiteColor];
    [[ExpoKit sharedInstance] application:application didFinishLaunchingWithOptions:launchOptions];
    _rootViewController = [ExpoKit sharedInstance].rootViewController;
    _window.rootViewController = _rootViewController;

    [_window makeKeyAndVisible];
    
    return YES;
}

#pragma mark - Handling URLs

- (BOOL)application:(UIApplication *)application openURL:(NSURL *)url sourceApplication:(nullable NSString *)sourceApplication annotation:(id)annotation
{
    return [[ExpoKit sharedInstance] application:application openURL:url sourceApplication:sourceApplication annotation:annotation];
}

- (BOOL)application:(UIApplication *)application continueUserActivity:(nonnull NSUserActivity *)userActivity restorationHandler:(nonnull void (^)(NSArray * _Nullable))restorationHandler
{
    return [[ExpoKit sharedInstance] application:application continueUserActivity:userActivity restorationHandler:restorationHandler];
}

#pragma mark - Notifications

// - (void)application:(UIApplication *)application didRegisterForRemoteNotificationsWithDeviceToken:(NSData *)token
// {
//     [[ExpoKit sharedInstance] application:application didRegisterForRemoteNotificationsWithDeviceToken:token];
// }

// - (void)application:(UIApplication *)application didFailToRegisterForRemoteNotificationsWithError:(NSError *)err
// {
//     [[ExpoKit sharedInstance] application:application didFailToRegisterForRemoteNotificationsWithError:err];
// }

// - (void)application:(UIApplication *)application didReceiveRemoteNotification:(NSDictionary *)notification
// {
//     [[ExpoKit sharedInstance] application:application didReceiveRemoteNotification:notification];
// }

// - (void)application:(UIApplication *)application didReceiveLocalNotification:(nonnull UILocalNotification *)notification
// {
//     [[ExpoKit sharedInstance] application:application didReceiveLocalNotification:notification];
// }

// - (void)application:(UIApplication *)application didRegisterUserNotificationSettings:(nonnull UIUserNotificationSettings *)notificationSettings
// {
//     [[ExpoKit sharedInstance] application:application didRegisterUserNotificationSettings:notificationSettings];
// }

#pragma mark - Firebase

- (void)application:(UIApplication *)application didReceiveLocalNotification:(UILocalNotification *)notification {
  [[RNFirebaseNotifications instance] didReceiveLocalNotification:notification];
}

- (void)application:(UIApplication *)application didReceiveRemoteNotification:(nonnull NSDictionary *)userInfo
                                                       fetchCompletionHandler:(nonnull void (^)(UIBackgroundFetchResult))completionHandler{
  [[RNFirebaseNotifications instance] didReceiveRemoteNotification:userInfo fetchCompletionHandler:completionHandler];
}

- (void)application:(UIApplication *)application didRegisterUserNotificationSettings:(UIUserNotificationSettings *)notificationSettings {
  [[RNFirebaseMessaging instance] didRegisterUserNotificationSettings:notificationSettings];
}

#pragma mark - FBSDK

- (BOOL)application:(UIApplication *)application
            openURL:(NSURL *)url
            options:(NSDictionary<UIApplicationOpenURLOptionsKey,id> *)options {
    
    BOOL handled = [[FBSDKApplicationDelegate sharedInstance] application:application
                                                                  openURL:url
                                                        sourceApplication:options[UIApplicationOpenURLOptionsSourceApplicationKey]
                                                               annotation:options[UIApplicationOpenURLOptionsAnnotationKey]
                    ];

    return handled;
}

- (void)applicationDidBecomeActive:(UIApplication *)application {
    [FBSDKAppEvents activateApp];
}

@end

```

</p>
</details>

---

### Android

<details><summary>Click To Expand</summary>
<p>

#### Have you converted to AndroidX?

- [ ] my application is an AndroidX application?
- [ ] I am using `android/gradle.settings` `jetifier=true` for Android compatibility?
- [ ] I am using the NPM package `jetifier` for react-native compatibility?

#### `android/build.gradle`:

```groovy
// N/A
```

#### `android/app/build.gradle`:

```groovy
// N/A
```

#### `android/settings.gradle`:

```groovy
// N/A
```

#### `MainApplication.java`:

```java
// N/A
```

#### `AndroidManifest.xml`:

```xml
<!-- N/A -->
```

</p>
</details>

---

## Environment

<details><summary>Click To Expand</summary>
<p>

**`react-native info` output:**

```
react-native info
Scanning folders for symlinks in /Users/a/mission/client/node_modules (37ms)

Environment:
  OS: macOS 10.15.1
  Node: 8.9.4
  Yarn: 1.16.0
  npm: 6.9.0
  Watchman: 4.9.0
  Xcode: Xcode 11.2.1 Build version 11B500
  Android Studio: Not Found

Packages: (wanted => installed)
  react: 16.3.1 => 16.3.1
  react-native: https://github.com/expo/react-native/archive/sdk-30.0.0.tar.gz => 0.55.4
```

- **Platform that you're experiencing the issue on**:
  - [x] iOS
  - [ ] Android
  - [ ] **iOS** but have not tested behavior on Android
  - [ ] **Android** but have not tested behavior on iOS
  - [ ] Both
- **`react-native-firebase` version you're using that has this issue:**
  - `5.0.0-rc4`
- **`Firebase` module(s) you're using that has the issue:**
  - `Auth`
- **Are you using `TypeScript`?**
  - `N`
  
</p>
</details>

---

Think `react-native-firebase` is great? Please consider supporting all of the project maintainers and contributors by donating via our [Open Collective](https://opencollective.com/react-native-firebase/donate) where all contributors can submit expenses. [[Learn More]](https://invertase.io/oss/react-native-firebase/contributing/donations-expenses)

- 👉 Check out [`React Native Firebase`](https://twitter.com/rnfirebase) and [`Invertase`](https://twitter.com/invertaseio) on Twitter for updates on the library.

**Comment by andersonaddo** [comment]
I know that it can be undefined (even if the user is authenticated) if the first onAuthStateChanged is not called (because the auth is still initializing).
It's stated in the [web docs](https://firebase.google.com/docs/auth/web/manage-users) and the [iOS docs](https://firebase.google.com/docs/auth/ios/manage-users).

**Comment by booboothefool** [comment]
@andersonaddo Hey, appreciate the input. I am already using that, however.

```
firebase.auth().onAuthStateChanged(function(user) {
  if (user) {
    // User is signed in. Only after this happens does my app redirect from the login and start making authenticated calls. This is what I meant when I said the user is authenticated, it still fails intermittently after this.
  } else {
    // No user is signed in. This always redirects back to the login screen.
  }
});
```

**Comment by andersonaddo** [comment]
Bump

**Comment by caxton** [comment]
bump

**Comment by mikehardy** [comment]
> bump

@caxton if you have a problem, open a new issue following the issue template. A bump on a closed issue has negative value, you clutter inboxes with notifications that are meaningless

**Comment by vesparny** [comment]
@booboothefool did you manage to solve it?

**Comment by adrianhorning08** [comment]
Same issue for me. I'm using firebase for users to sign into a chrome extension. Can't replicate on my machine. And it sometimes works for users, sometimes not.

**Comment by mikehardy** [comment]
If it is a chrome extension, how can it be react-native :thinking: ? wouldn't that be firebase-js-sdk ?
